[ROCm] hipblaslt integration

zstreet87 opened the pull request:
This PR adds hipblaslt capabilities.  

Hipblaslt will be available in ROCm5.5.  Also, this PR requires the user to opt-in to using the hipblaslt path with the env var `USE_HIPBLASLT=1`. 

Any suggestions for a cleaner integration design is more than welcomed but we followed the cublaslt design.  A better design will require hipify to know about hiblaslt datatypes. This PR provides a workaround until that work is done (i.e. hipify is modified).  

Requires https://github.com/pytorch/pytorch/pull/103092

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport

rraminen committed:
Skipped the tests as the fail w.r.t https://github.com/ROCmSoftwarePlatform/pytorch/pull/993

rraminen committed:
[SOW MS3] Centos stream9 PyTorch image support (#1090)

* changes to build Centos stream 9 images

* Added scripts for centos and centos stream images

* Added an extra line

* Add ninja installation

* Optimized code

* Fixes

* Add comment

* Optimized code

* Added AMDGPU mapping for ROCm 5.2 and invalid-url for rocm_baseurl

Co-authored-by: Jithun Nair <37884920+jithunnair-amd@users.noreply.github.com>

pruthvistony committed:
Updated to latest conda for CentOS stream 9

jataylo committed:
Add test_jit_cuda_fuser to ROCM_BLOCKLIST (#1139)

pruthvistony committed:
Enable tensorpipe with hip_basic backend (#1135)

* Add hip_basic tensorpipe support to PyTorch

* Enabling hip_basic for Tensorpipe for pyTorch

* removing upstream tensorpipe module

* Adding ROCm specific tensopipe submodule

* tensorpipe submodule updated

* Update the hip invalid device string

* Added ignore for tensorpipe git submodule

* Moved include of tensorpipe_cuda.h to hipify

* Updates based on review comments

* Defining the variable __HIP_PLATFORM_AMD__

* Enabling the UTs

Co-authored-by: Ronak Malik <Ronak.Malik@amd.com>

BLOrange-AMD committed:
Update for new MIOpen kernel package names (#1140)

jithunnair-amd committed:
also look for 'Cijk' to verify gemm in test_kineto

alugorey committed:
Pull in hipSOLVER code from sow_ms7 (#1156)

* [SOW MS3] hipSOLVER integration (#1036)

* [ROCM] Add cuSOLVER hipify mappings.

Similar to cuSPARSE, cuSOLVER APIs need different data types to
translate cuComplex, cuDoubleComplex, etc. into.

Note: is_cusolver_file only checks if the path contains "linalg".

* [ROCM] Workaround for SWDEV-344283

This commit should be reverted once the hipSOLVER bug was fixed.

* [ROCM][Linalg] Enable functions that are already supported by ROCM.

This requires ROCM version >= 5.2

* [ROCM] enable tests for cuSOLVER on ROCM >= 5.2

* (to squash) remove extra empty lines

* (to squash) remove the unnecessary include of rocsolver.h

* Make TORCH_CUSOLVER_CHECK available for ROCM >= 5.20

* CUDAHooks::hasCuDNN() returns true for ROCM_VERSION >= 50200

Fix "RuntimeError: Cannot set preferred backend to cuSOLVER if PyTorch
has not been compiled with cuSOLVER." in torch._C._set_linalg_preferred_backend

* [ROCM] enable hipSOLVER backend at svd_kernel()

Note on ROCM the only gesvd kernel is enabled for svd.
gesvdj and gesvda are disabled for now.

* Only enable the following cases in test_linalg

test_householder_product
test_householder_product_errors_and_warnings
test_ormqr

* [ROCM] unskip test_ormqr_errors_and_warnings

* (fix) [ROCM] enable hipSOLVER backend at svd_kernel()

* Revert "[ROCM] enable hipSOLVER backend at svd_kernel()"

This reverts commit f0849574caee30a9b94f526319442d262be7582f.

* Revert "(fix) [ROCM] enable hipSOLVER backend at svd_kernel()"

This reverts commit 7f54a81ac97f51aa007609f9b26752e9b17293b7.

* Revert the enablement of TORCH_CUSOLVER_CHECK

* Revert "CUDAHooks::hasCuDNN() returns true for ROCM_VERSION >= 50200"

This reverts commit 95e022281895d336b612dbbd0b277bb19c36650a.

* [SOW MS3] Enable test_svd_memory_allocation (#1060)

* [ROCM] Enable hipSOLVER backend for torch.linalg.svd

This is essential for test_svd_memory_allocation, which is not supported
on Magma.

* [ROCM] svd_kernel workaround 1: force using gesvd

gesvdj and gesvda are not fully implemented in rocSOVLER.

* [ROCM] svd_kernel workaround 2: handling the memory layout on ROCM.

It seems hipSOLVER's gesvd handles m < n in a slightly different way with cuSOLVER.

* [ROCM] Enable test_svd_memory_allocation on ROCM

* Enable torch.backends.cuda.preferred_linalg_library

* Bump the hipSOLVER requirement to ROCM 5.3

* Also test 'cusolver' backend when hipSOLVER is available

* Fix RuntimeError torch.linalg.svd: keyword argument `driver=` is only supported on CUDA inputs with cuSOLVER backend.

* Move hasHIPSOLVER from HIPHooksInterface to CUDAHooksInterface

* [ROCM] More workaround on svd_kernel

1. hipSOVLER's gesvd has poor performance on large sqaure matrices.
   Fall back to magma in this case.
2. Ensure Vh is not guarateed to fit svdMagma requirements.

* [ROCM] Better way to handle Vh memory in svd_cusolver_gesvd

Do not copy, use at::empty_like.

* [ROCM] remove @skipCUDAIfRocm from test_svd_memory_allocation

* [ROCM] Document storage order handling of Vh output tensor in svd related functions

* [ROCM] set precision_overrides when testing 'nuc' norm because it is based on SVD.

* Re-add CONST_QUALIFIER for rocm version less than 5.3 (#1088)

* Fix CONST_QUALIFIER definition (#1089)

* Set magma backend for meta tests (#1087)

* Set magma backend for test_out/warning (#1097)

* [SOW MS3] Enable gesvdj API to be called (#1091)

* Enable gesvdj API to be called

* Force driver to gesvdj

* Set driver based on arch

* Remove duplicate cuda to hip mappings

* [SOW MS3] Enable hipSOLVER for test_ops*

Updated cherry-pick of PR#1061
- Use skipCUDAIfNoCusolverAndNoHipsolver replacing
   skipCUDAIfNoCusolver only for decorators list of:
  - linalg.householder_product
  - ormqr

* Bump ROCm version requirements to at least 5.3

Co-authored-by: Xinya Zhang <xinyazhang@utexas.edu>
Co-authored-by: Pruthvi Madugundu <pmagundu@amd.com>

jithunnair-amd committed:
Only install miopen-hip-gfx packages for ROCm5.5 onwards (#1164)

* Starting from ROCm5.5, miopenkernels packages are replaced by miopen-hip-gfx packages

* Exit if miopen packages not available

* Exit if miopen packages not available

ashutom committed:
* [ROCM] making cuSOLVER public (#1170)

* [ROCM] Solution for SWDEV-379692

zstreet87 committed:
initial commit integration of hipblaslt before code clean-up

zstreet87 committed:
added ROCM_VERSION checks, need to test on ROCm5.5 image

zstreet87 committed:
Merge branch 'pytorch:master' into hipblaslt_integration

zstreet87 committed:
Merge branch 'pytorch:master' into hipblaslt_integration

zstreet87 committed:
Merge branch 'pytorch:master' into hipblaslt_integration

zstreet87 committed:
USE_ROCM for the exception files

zstreet87 committed:
Merge branch 'hipblaslt_integration' of https://github.com/zstreet87/pytorch into hipblaslt_integration

zstreet87 committed:
changed macro name

Skylion007 commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -263,6 +267,12 @@ void bgemm<at::Half>(CUDABLAS_BGEMM_ARGTYPES(at::Half)) {
                                    (int) num_batches, rocblas_datatype_f32_r, rocblas_gemm_algo_standard,
                                    0, flag));
 #else
+  #if defined(CUDA_VERSION) && CUDA_VERSION < 11000
```
We don't support pre-C++11 CUDA anymore.

zstreet87 committed:
removed cuda < 11000 and added case statement syntax

alugorey committed:
Integrate new batched linalg drivers (#1163)

* Integrate new batched linalg drivers

* Skip test_qr_batched; ROCM doesn't support QR decomp for complex dtype

* Skip complex types, hipsolver does not support

* Skip complex types in other batched tests as well

pruthvistony committed:
Adding a 64GB limit for cross_entropy_large test

- https://ontrack-internal.amd.com/browse/SWDEV-373709
  ROCm needs more memory and no GPU support with >64GB,
  so keeping this change on ROCm fork only.

pruthvistony committed:
Skipping test_complex_half_reference_testing for conv 3D operator

pruthvistony committed:
Skip all inductor related UT

- Triton not supported in rocm5.5_internal_testing branch

wenchenvincent committed:
Sync updates from hipify_torch. (#1168)

Co-authored-by: Jithun Nair <37884920+jithunnair-amd@users.noreply.github.com>

pruthvistony committed:
Add CI=1 environment variable as part of test.sh (#1198)

jithunnair-amd committed:
Add CHANGELOG.md stub

pruthvistony committed:
Temporarily skip test_conv3d_64bit_indexing

- Rocblas API support is requested
- SWDEV-383635 & sub task - SWDEV-390218

pruthvistony committed:
[ROCm] Remove usage of deprecated ROCm component includes (#1207)

* [ROCm] Remove usage of deprecated HIP includes

* Review comment update

* Roctracer header include path updated

pruthvistony committed:
Moving gloo submodule to ROCm fork

- branch rocm5.6_internal_testing

Updated the gloo submodule

jithunnair-amd committed:
Set CONDA_CMAKE=yes for any 'rocm' configs (#1210)

BLOrange-AMD committed:
Fixed test_memory_timeline in 5.6 (#1204)

zstreet87 committed:
Merge branch 'main' into hipblaslt_integration

jeffdaily committed:
conditionally enable hipsparse const descriptors for version >= 2.4.0 (#1217)

* conditionally enable hipsparse const descriptors

* update hipsparse const API version condition to 2.4.0

zstreet87 committed:
Merge branch 'pytorch:main' into hipblaslt_integration

pruthvistony committed:
Limiting the num_procs to 8 (#1219)

- SWDEV-385909
- Inturns limits the NUM_PARALLEL_PROCS to 8, so the test
  shards are also max 8

zstreet87 committed:
added USE_ROCM guards for new int8 and alignment

jithunnair-amd committed:
Add related_commits (#1220)

* Add related_commits with apex commit updated; need to update other torch domain library commits

* Update torch domain library commits

BLOrange-AMD committed:
Modified logging threshold for memory profiling (#1227)

zstreet87 committed:
Merge branch 'main' into hipblaslt_integration

pruthvistony committed:
Create .jenkins symlink for backward compatibility

- .jenkins ->.ci
- Remove .jenkins reference from devops scripts
- Remove .jenkins reference from test scrips

pruthvistony committed:
Revert of [ci] remove old USER_LAND code from ci scripts

- https://github.com/pytorch/pytorch/pull/80045

pruthvistony committed:
Disabling NVFUSER by default

pruthvistony committed:
[ROCm] Fix test_nadam

- Cherry-pick of 5fc34231aba91b2397c25a9fa5ce0d97ae11099c

pruthvistony committed:
Remove wrong install_magma

pruthvistony committed:
Remove error hip mapping for cublasSgetriBatched

pruthvistony committed:
Revert "Create .jenkins symlink for backward compatibility"

This reverts commit 798f1782ffa52103c62b075b2ebc7f44ac205c36.

xinyazhang committed:
[ROCM] Fix TestLinalgCUDA.test_qr_cuda_complex64.

xinyazhang committed:
ROCM: enable hipSOLVER backend for ldl_factor

skc7 committed:
[LLVM] Make changes needed for opaque pointers (#101396)

Update llvm_codegen module to use opaque pointers feature of llvm.

* Set setOpaquePointers to true for llvm context.
* Pass Type to emit\*Load and emit\*Store functions.
* Create TypedPointer struct to keep track of Value and its Type.
* Introduce OpqTy_ to be used for opaque pointer types.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/101396
Approved by: https://github.com/jgong5

zstreet87 committed:
Merge branch 'main' into hipblaslt_integration

pruthvistony committed:
Merge pull request #1236 from ROCmSoftwarePlatform/IFU-main-2023-06-06

IFU-main-2023-06-06

pruthvistony committed:
Merge branch 'downstream_main' into rocm5.7_internal_testing

pruthvistony committed:
Updated related_commits file

jeffdaily committed:
fix merge commit 006b3e67c188b7b0d27b543e9536076d6b389c85

combines hipSOLVER and hipSPARSE mappings into single SPECIAL map that takes priority among normal mappings

the constant API_SPARSE was renamed API_SPECIAL and combined with API_SOLVER

lcskrishna committed:
[SWDEV-396381] FSDP UT failures (#1234)

* use 8 gpus on rocm if a node has more than 8 gpus

* limit to 8 gpus for FSDP tests

* add message in change log for fsdp-swdev396381

jithunnair-amd committed:
Update CHANGELOG.md

zstreet87 committed:
Merge branch 'pytorch:main' into hipblaslt_integration

zstreet87 committed:
Merge branch 'rocm5.7_internal_testing' into hipblaslt_integration
